File: dsalgo/leetcode/easy/isAlienSorted.py

def isAlienSorted(words, order):
    """
    :type words: List[str]
    :type order: str
    :rtype: bool
    """
    ind = {c: i for i, c in enumerate(order)}
    for a, b in zip(words, words[1:]):
        if len(a) > len(b) and a[:len(b)] == b:
            return False
        for s1, s2 in zip(a, b):
            if ind[s1] < ind[s2]:
                break
            elif ind[s1] > ind[s2]:
                return False
    return True
    
#similar solution
def isAlienSorted(words, order):
    """
    :type words: List[str]
    :type order: str
    :rtype: bool
    """
    order = {c: i for i, c in enumerate(order)}

    for w1, w2 in zip(words, words[1:]):
        for c1, c2 in zip(w1, w2):
            if c1 != c2:
                if order[c1] > order[c2]:
                    return False
                break
        else:
            if len(w2) < len(w1):
                return False

    return True

print(isAlienSorted(["my", "f"], "gelyriwxzdupkjctbfnqmsavho"))

File: dsalgo/leetcode/easy/isAlienSorted.js



// const isAlienSorted = function(words, order) {

// };

// const compare = function(s1, s2, order) {
//     if (s1 === s2) return true;


// }
var isAlienSorted = function(words, order) {
    for (let i = 0; i < words.length - 1; i++) {
        const prev = words[i];
        const next = words[i + 1];
        let j = 0;
        while (prev[j] && next[j] || !(prev[j] && next[j])) {
            next[j] = next[j] ? next[j] : -1;
            if ((order.indexOf(next[j]) - order.indexOf(prev[j])) > 0) {
                return true;
            }
            if ((order.indexOf(next[j]) - order.indexOf(prev[j])) < 0) {
                return false;
            }
            if (next[j] === undefined && prev[j] === undefined) {
                return true;
            }
            j++;
        }
    }
};

// const isAlienSorted = (words, order) => {
//     let obj = {};
//     for (let i = 0; i < order.length; i++) {
//         obj[order[i]] = i;
//     }

//     for (let i = 0; i < words.length - 1; i++) {
//         const word = words[i];
//         const next = words[i + 1];
//         let pointer = 0;
//         while(word[pointer] === next[pointer]) pointer++;
//         if (next[pointer] === undefined || obj[word[pointer] > obj[next[pointer]]]) {
//             return false;
//         }
//     }
//     return true;
// }

console.log(isAlienSorted(["fxasxpc","dfbdrifhp","nwzgs","cmwqriv","ebulyfyve","miracx","sxckdwzv","dtijzluhts","wwbmnge","qmjwymmyox"], "zkgwaverfimqxbnctdplsjyohu")); // false
// console.log(isAlienSorted(["my", "f"], "gelyriwxzdupkjctbfnqmsavho")); // false
// console.log(isAlienSorted(["hello", "hello"], "abcdefghijklmnopqrstuvwxyz")); // true
// console.log(isAlienSorted(["hello","leetcode"], "hlabcdefgijkmnopqrstuvwxyz")); // true
// console.log(isAlienSorted(["word","world","row"], "worldabcefghijkmnpqstuvxyz")); // false
// console.log(isAlienSorted(["apple","app"], "abcdefghijklmnopqrstuvwxyz")); // false
// console.log(isAlienSorted(["kuvp","q"], "ngxlkthsjuoqcpavbfdermiywz")); // true
